matlab/CMakeLists.txt

```diff
@@ -317,6 +324,51 @@ matlab_add_mex(R2018a
 
 target_include_directories(mexcall PRIVATE ${CPP_SOURCE_DIR})
 
+# ARROW_SHARED_LIB
+# On Windows, this will be ARROW_HOME/bin/arrow.dll and on Linux and macOS, it is the arrow.so/dylib in the newly built arrow_shared library.
+if(NOT Arrow_FOUND)
+  message(STATUS "ARROW_SHARED_LIB will be set using IMPORTED_LOCATION value when building."
+  )
+  get_target_property(ARROW_SHARED_LIB arrow_shared IMPORTED_LOCATION)
+else()
+  # If not building Arrow, ARROW_SHARED_LIB derived from ARROW_PREFIX set to the ARROW_HOME specified with cmake would be non-empty.
+  message(STATUS "ARROW_SHARED_LIB: ${ARROW_SHARED_LIB}")
+endif()
+
+# ARROW_LINK_LIB
+# On Windows, we use the arrow.lib for linking arrow_matlab against the Arrow C++ library.
+# The location of arrow.lib is previously saved in IMPORTED_IMPLIB.
+if(WIN32)
+  # If not building Arrow, IMPORTED_IMPLIB will be empty.
+  # Then set ARROW_LINK_LIB to ARROW_IMPORT_LIB which would have been derived from ARROW_PREFIX set to the ARROW_HOME specified with cmake. This will avoid the ARROW_LINK_LIB set to NOTFOUND error.
+  # The ARROW_IMPORT_LIB should be ARROW_HOME/lib/arrow.lib on Windows.
+  if(NOT Arrow_FOUND)
+    message(STATUS "ARROW_LINK_LIB will be set using IMPORTED_IMPLIB value when building."
+    )
+    get_target_property(ARROW_LINK_LIB arrow_shared IMPORTED_IMPLIB)
+  else()
+    set(ARROW_LINK_LIB "${ARROW_IMPORT_LIB}")
+    message(STATUS "Setting ARROW_LINK_LIB to ARROW_IMPORT_LIB: ${ARROW_IMPORT_LIB}, which is derived from the ARROW_HOME provided."
+    )
+  endif()
+else()
+  # On Linux and macOS, it is the arrow.so/dylib in the newly built arrow_shared library used for linking.
+  # On Unix, this is the same as ARROW_SHARED_LIB.
+  message(STATUS "Setting ARROW_LINK_LIB to ARROW_SHARED_LIB as they are same on Unix.")
+  set(ARROW_LINK_LIB "${ARROW_SHARED_LIB}")
+endif()
+
+# ARROW_INCLUDE_DIR should be set so that header files in the include directory can be found correctly.
+# The value of ARROW_INCLUDE_DIR should be ARROW_HOME/include on all platforms.
+if(NOT Arrow_FOUND)
+  message(STATUS "ARROW_INCLUDE_DIR will be set using INTERFACE_INCLUDE_DIRECTORIES value when building."
+  )
+  get_target_property(ARROW_INCLUDE_DIR arrow_shared INTERFACE_INCLUDE_DIRECTORIES)
+else()
+  # If not building Arrow, ARROW_INCLUDE_DIR derived from ARROW_PREFIX set to the ARROW_HOME specified with cmake would be non-empty.
+  message(STATUS "ARROW_INCLUDE_DIR: ${ARROW_INCLUDE_DIR}")
+endif()
+
```

Comment:
Could you use the `Arrow::arrow_shared` CMake target (or `Arrow::arrow_static`) instead of separated library path, include path and so on?
The `Arrow::arrow_shared` CMake target has all information (library path, include path and so on) we need.

Reply:
Thanks for the suggestion, @kou!

Just to clarify, are you suggesting that we should use `Arrow::arrow_shared` in the `else` clauses where `Arrow_FOUND` is `true` for printing debug messages? Or are you suggesting, more generally, that we don't set imported target properties separately on `arrow_shared`?

For reference, when building Arrow from source using `ExternalProject_Add` (i.e. when `find_package(Arrow)` fails to find a valid Arrow installation), the imported target that is created from the `ExternalProject_Add` build artifacts is called `arrow_shared`. If we start using `Arrow::arrow_shared`, there would then be two possible different targets: (1) `Arrow::arrow_shared` and (2) `arrow_shared`, which might complicate the logic of the build system (although, that doesn't necessarily mean we shouldn't make this change - but, it seems like a consideration worth keeping in mind).

Reply:
Ah, sorry. I misread the implementation.
I thought all individual target properties are used directly for building a library because https://github.com/apache/arrow/pull/34563/files#diff-282bdaf7afd3f3d7974e3ab41857a65e3eea6a566107ea7d209ba3fec72e2e77R43 uses `ARROW_INCLUDE_DIRS`.
But I noticed that `ARROW_INCLUDE_DIRS` is only used directly and we can remove it because `arrow_shared` is specified: https://github.com/apache/arrow/pull/34563/files#diff-282bdaf7afd3f3d7974e3ab41857a65e3eea6a566107ea7d209ba3fec72e2e77R81

FYI: `arrow_shared` is deprecated in the current `find_package(Arrow)`. `Arrow::arrow_shred` is recommended. We can use `Arrow::arrow_shared` instead of `arrow_shared` for `ExternalProject_Add()`-ed Apache Arrow C++. But it'll be out-of-scope of this pull request.

Reply:
No problem at all! Admittedly, the logic is a bit confusing since we currently support building from source.

Ideally, it would be great if we could refactor the build system in the future to use `FetchContent` when building from source (instead of `ExternalProject_Add`). However, it seems like `FetchContent` isn't currently supported based on #32234.

Thanks for letting us know that `arrow_shared` is deprecated. We agree that we should eventually move to using `Arrow::arrow_shared`. This is something that we can work on for a future pull request.

Reply:
I have captured this in #35309. We can probably mark this as resolved now.